Casting 0xFFFFFFFF to handle generates a warning in MSVC

### Background

In src/platform/win32/WinIOCallback.cpp, a file handle is created using CreateFileA. We check that the result is valid before we try to use the handle as such:
```
  mFile = CreateFileA(Path, AccessMode, ShareMode, NULL, Disposition, dwFlags, NULL);
  if ((mFile == INVALID_HANDLE_VALUE) || (mFile == (HANDLE)0xffffffff)) {
```

### Bug
MSVC does not like the casting of 0xFFFFFFFF to type handle. This generates the warning 
`warning C4312: 'type cast': conversion from 'unsigned int' to 'HANDLE' of greater size`

### Open question: Is this also a bug?
We check that the handle is valid by comparing it to (I assume) -1, written as 0xFFFFFFFF. According to the [Microsoft documentation for CreateFileA](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea), the two return values are either a valid handle, or INVALID_HANDLE_VALUE. Why do we also check for 0xFFFFFFFF?

### Proposed fix
If the 0xFFFFFFFF check is unnecessary, we can simply remove that check. If it is necessary, we'll need to update the value so the cast can be done safely/correctly.

--- a/.github/workflows/windows.yaml
+++ b/.github/workflows/windows.yaml
@@ -7,7 +7,7 @@ on:
 
 jobs:
   build_libebml:
-    name: libebml ${{ matrix.config }} ${{ matrix.shared }} Library
+    name: libebml ${{ matrix.config }} ${{ matrix.shared }}
     runs-on: windows-latest
     strategy:
       fail-fast: false
@@ -24,11 +24,11 @@ jobs:
       
       - name: Configure static library
         if: ${{ matrix.shared == 'Static' }}
-        run: cmake -S . -B _build ${{ env.CMAKE_OPTIONS }} -DCMAKE_INSTALL_PREFIX:STRING=${GITHUB_WORKSPACE}/_built -DBUILD_SHARED_LIBS=OFF
+        run: cmake -S . -B _build ${{ env.CMAKE_OPTIONS }} -DBUILD_SHARED_LIBS=OFF
 
       - name: Configure dynamic library
         if: ${{ matrix.shared == 'Dynamic' }}
-        run: cmake -S . -B _build ${{ env.CMAKE_OPTIONS }} -DCMAKE_INSTALL_PREFIX:STRING=${GITHUB_WORKSPACE}/_built -DBUILD_SHARED_LIBS=ON
+        run: cmake -S . -B _build ${{ env.CMAKE_OPTIONS }} -DBUILD_SHARED_LIBS=ON
 
       - name: Build
         run: cmake --build _build --config ${{ matrix.config }} --parallel

--- a/.github/workflows/windows.yaml
+++ b/.github/workflows/windows.yaml
@@ -14,6 +14,8 @@ jobs:
       matrix:
         config: [Debug, Release]
         shared: [Dynamic, Static]
+    env:
+      CMAKE_OPTIONS: -DENABLE_WIN32_IO=ON
     steps:
       - uses: lukka/get-cmake@latest
 
@@ -22,11 +24,11 @@ jobs:
       
       - name: Configure static library
         if: ${{ matrix.shared == 'Static' }}
-        run: cmake -S . -B _build -DCMAKE_INSTALL_PREFIX:STRING=${GITHUB_WORKSPACE}/_built -DBUILD_SHARED_LIBS=OFF
+        run: cmake -S . -B _build ${{ env.CMAKE_OPTIONS }} -DCMAKE_INSTALL_PREFIX:STRING=${GITHUB_WORKSPACE}/_built -DBUILD_SHARED_LIBS=OFF
 
       - name: Configure dynamic library
         if: ${{ matrix.shared == 'Dynamic' }}
-        run: cmake -S . -B _build -DCMAKE_INSTALL_PREFIX:STRING=${GITHUB_WORKSPACE}/_built -DBUILD_SHARED_LIBS=ON
+        run: cmake -S . -B _build ${{ env.CMAKE_OPTIONS }} -DCMAKE_INSTALL_PREFIX:STRING=${GITHUB_WORKSPACE}/_built -DBUILD_SHARED_LIBS=ON
 
       - name: Build
         run: cmake --build _build --config ${{ matrix.config }} --parallel

--- a/.github/workflows/windows.yaml
+++ b/.github/workflows/windows.yaml
@@ -7,13 +7,16 @@ on:
 
 jobs:
   build_libebml:
-    name: libebml ${{ matrix.config }} ${{ matrix.shared }}
+    name: libebml ${{ matrix.config }} ${{ matrix.shared.name }}
     runs-on: windows-latest
     strategy:
       fail-fast: false
       matrix:
         config: [Debug, Release]
-        shared: [Dynamic, Static]
+        shared: [
+          { "name": "Dynamic", "option": "ON"},
+          { "name": "Static",  "option": "OFF"}
+        ]
     env:
       CMAKE_OPTIONS: -DENABLE_WIN32_IO=ON
     steps:
@@ -22,13 +25,8 @@ jobs:
       - name: Get pushed code
         uses: actions/checkout@v3
       
-      - name: Configure static library
-        if: ${{ matrix.shared == 'Static' }}
-        run: cmake -S . -B _build ${{ env.CMAKE_OPTIONS }} -DBUILD_SHARED_LIBS=OFF
-
-      - name: Configure dynamic library
-        if: ${{ matrix.shared == 'Dynamic' }}
-        run: cmake -S . -B _build ${{ env.CMAKE_OPTIONS }} -DBUILD_SHARED_LIBS=ON
+      - name: Configure ${{ matrix.shared.name }} library
+        run: cmake -S . -B _build ${{ env.CMAKE_OPTIONS }} -DBUILD_SHARED_LIBS=${{ matrix.shared.option }}
 
       - name: Build
         run: cmake --build _build --config ${{ matrix.config }} --parallel

--- a/src/platform/win32/WinIOCallback.cpp
+++ b/src/platform/win32/WinIOCallback.cpp
@@ -95,7 +95,7 @@ bool WinIOCallback::open(const char* Path, const open_mode aMode, DWORD dwFlags)
   }
 
   mFile = CreateFileA(Path, AccessMode, ShareMode, NULL, Disposition, dwFlags, NULL);
-  if ((mFile == INVALID_HANDLE_VALUE) || (mFile == (HANDLE)0xffffffff)) {
+  if (mFile == INVALID_HANDLE_VALUE) {
     //File was not opened
     char err_msg[256];
     DWORD error_code = GetLastError();
@@ -177,7 +177,7 @@ bool WinIOCallback::open(const wchar_t* Path, const open_mode aMode, DWORD dwFla
       return mOk = false;
     }
   }
-  if ((mFile == INVALID_HANDLE_VALUE) || (mFile == (HANDLE)0xffffffff)) {
+  if (mFile == INVALID_HANDLE_VALUE) {
     //File was not opened
     char err_msg[256];
     DWORD error_code = GetLastError();

--- a/src/platform/win32/WinIOCallback.cpp
+++ b/src/platform/win32/WinIOCallback.cpp
@@ -146,37 +146,7 @@ bool WinIOCallback::open(const wchar_t* Path, const open_mode aMode, DWORD dwFla
       assert(false);
   }
 
-  if ((LONG)GetVersion() >= 0) {
-    mFile = CreateFileW(Path, AccessMode, ShareMode, NULL, Disposition, dwFlags, NULL);
-  } else {
-    int errCode;
-    int pathSize = wcslen(Path);
-    unsigned int bufferSize = pathSize + sizeof(wchar_t) * 2;
-    std::string PathA;
-    PathA.resize(bufferSize);
-    errCode = WideCharToMultiByte(CP_ACP, 0, Path, pathSize, (char *)PathA.c_str(), bufferSize, NULL, NULL);
-    if (errCode == 0)
-      errCode = GetLastError();
-#ifdef _DEBUG
-    if (errCode == ERROR_INSUFFICIENT_BUFFER) OutputDebugString(TEXT("WinIOCallback::WideCharToMultiByte::ERROR_INSUFFICIENT_BUFFER"));
-    if (errCode == ERROR_INVALID_FLAGS) OutputDebugString(TEXT("WinIOCallback::WideCharToMultiByte::ERROR_INVALID_FLAGS"));
-    if (errCode == ERROR_INVALID_PARAMETER) OutputDebugString(TEXT("WinIOCallback::WideCharToMultiByte::ERROR_INVALID_PARAMETER"));
-#endif
-    while (errCode == ERROR_INSUFFICIENT_BUFFER) {
-      // Increase the buffer size
-      bufferSize += MAX_PATH;
-      PathA.resize(bufferSize);
-      errCode = WideCharToMultiByte(CP_ACP, WC_SEPCHARS, Path, pathSize, (char *)PathA.c_str(), bufferSize, NULL, NULL);
-      if (errCode == 0)
-        errCode = GetLastError();
-    }
-    if (errCode != 0) {
-      mFile = CreateFileA(PathA.c_str(), AccessMode, ShareMode, NULL, Disposition, dwFlags, NULL);
-    } else {
-      mLastErrorStr = "Couldn't convert Unicode filename to ANSI.";
-      return mOk = false;
-    }
-  }
+  mFile = CreateFileW(Path, AccessMode, ShareMode, NULL, Disposition, dwFlags, NULL);
   if (mFile == INVALID_HANDLE_VALUE) {
     //File was not opened
     char err_msg[256];

--- a/src/platform/win32/WinIOCallback.cpp
+++ b/src/platform/win32/WinIOCallback.cpp
@@ -248,7 +248,7 @@ void WinIOCallback::setFilePointer(std::int64_t Offset, seek_mode Mode)
   }
 }
 
-std::uint32_t WinIOCallback::read(void*Buffer,std::size_t Size)
+std::size_t WinIOCallback::read(void*Buffer,std::size_t Size)
 {
   DWORD BytesRead;
   if (!ReadFile(mFile, Buffer, Size, &BytesRead, NULL)) {

--- a/src/platform/win32/WinIOCallback.h
+++ b/src/platform/win32/WinIOCallback.h
@@ -53,7 +53,7 @@ public:
   bool open(const wchar_t* Path, const open_mode Mode, DWORD dwFlags=0);
   bool open(const char* Path, const open_mode Mode, DWORD dwFlags=0);
 
-  virtual std::uint32_t read(void*Buffer,std::size_t Size);
+  virtual std::size_t read(void*Buffer,std::size_t Size);
   virtual std::size_t write(const void*Buffer,std::size_t Size);
   virtual void setFilePointer(std::int64_t Offset,seek_mode Mode=seek_beginning);
   virtual std::uint64_t getFilePointer();

